### PR TITLE
fixes #1027; disallow templates to use ambiguous identifiers

### LINKED ISSUE #1027: Templates allowed to use ambiguous identifier
The following set of files defines two modules with the same constant and a third one trying to `echo` it:

``` nimrod
# a.nim
const version_str* = "mod a"
# b.nim
const version_str* = "mod b"
# base.nim
import a, b

echo version_str
```

Compilation of the `base.nim` file fails correctly with:

```
base.nim(4, 5) Error: ambiguous identifier: 'version_str' -- use a qualifier
```

Wrapping the ambiguous identifier in a template avoids the error:

``` nimrod
# base.nim
import a, b

template wrap_me(stuff): stmt =
  echo "Using " & version_str

wrap_me("hey")
```

This executes using the symbol defined in `a.nim`:

```
$ nimrod c -r base
Using mod a
```

### PR DESCRIPTION
**Pull request by ringabout**
fixes https://github.com/nim-lang/Nim/issues/1027

### CODE REVIEW

**Comment by bung87** [comment]
doesn't this root cause is by ?

``` nim
let candidates = searchInScopesFilterBy(c, ident, allExceptModule) #.skipAlias(n, c.config)
if candidates.len > 0:
  result = candidates[0]
  amb = candidates.len > 1
  if amb and checkAmbiguity in flags:
    errorUseQualifier(c, n.info, candidates) 
```

after latest refactoring of  `searchInScopesFilterBy`, `candidates.len > 1` never will be true, which means `checkAmbiguity` become useless.

**Comment by ringabout** [comment]
> after latest refactoring of searchInScopesFilterBy, candidates.len > 1 never will be true, which means checkAmbiguity become useless.

Where is it?

**Comment by bung87** [comment]
> > after latest refactoring of searchInScopesFilterBy, candidates.len > 1 never will be true, which means checkAmbiguity become useless.
> 
> Where is it?

https://github.com/nim-lang/Nim/commit/8f4bdb3596bb7650f2688b8243dc3164abfc21c9

**Comment by ringabout** [comment]
> https://github.com/nim-lang/Nim/commit/8f4bdb3596bb7650f2688b8243dc3164abfc21c9

I don't think it is related. That PR is regarding shadowing variables.

**Comment by bung87** [comment]
> > [8f4bdb3](https://github.com/nim-lang/Nim/commit/8f4bdb3596bb7650f2688b8243dc3164abfc21c9)
> 
> I don't think it is related. That PR is regarding shadowing variables.

really? so how can `candidates.len > 1 ` become true ?

**Comment by ringabout** [comment]
My PR [8f4bdb3](https://github.com/nim-lang/Nim/commit/8f4bdb3596bb7650f2688b8243dc3164abfc21c9) didn't change the correctness but the performance, two pieces of code are essentially identical.

**Comment by bung87** [comment]
so maybe it doesn't work properly in previous ?

File: tests/template/m1027a.nim [hunk 1]
```diff
@@ -0,0 +1 @@
+const version_str* = "mod a"
```
Comment (Varriount) [inline tests/template/m1027a.nim]:
```suggestion
const version_str* = "mod a"

```

File: tests/template/m1027b.nim [hunk 2]
```diff
@@ -0,0 +1 @@
+const version_str* = "mod b"
```
Comment (Varriount) [inline tests/template/m1027b.nim]:
```suggestion
const version_str* = "mod b"

```

**Comment by tersec** [comment]
Would be useful to have this backported to 1.6 -- it silently generates invalid code.

**Comment by arnetheduck** [comment]
cc @narimiran

**Pull request merged**